create_layout.py

```diff
@@ -91,45 +91,111 @@
 import in_toto.models.link
 import in_toto.models.layout
 
-def create_material_rules(links, index):
+def changes_between_snapshots(before_dict, after_dict):
+  '''Given two 'snapshots' of a file structure -- 'before' and 'after' -- 
+    return a tuple specifying which files have been added, 
+    which have been removed, which have been modified, 
+    and which have remained unchanged. 
+    Both these dictionaries have file
+    names as the keys and their hashes as the values.'''
+
+  unchanged_files = set()
```

Comment:
So, I think I understand why this was updated to use sets, but I think it's worth noting that `sorted` below converts it back into a list. So I only see a reason to use sets if we want to get rid of the multiple `for` loops.

Reply:
I do want to emphasize that you don't have to do it this way, but what I meant when I mentioned sets here is something along the lines of:

```
before_set = set(before_dict.keys())
after_set = set(after_dict.keys())

removed_keys = sorted(before_set.difference(after_set))
added_keys = sorted(after_set.difference(before_set))

for key in before_set.intersection(after_set)):
  if before_dict[key] == after_dict[key]:
    unchanged_files.append(key) # sort later
  else:
    modified_files.append(key) # sort later
```

Reply:
The only reason we return sorted list here is for the unittest. If we don't sort the list, `changes_between_snapshots` will return a list in a random order. As a result, `assertEqual` will occasionally return false unexpectedly.
Isha wrote this as a quick fix to the problem above, but I don't think it is necessary to sort the list here. Maybe we can just return sets, and use `assertSetEqual`. Or maybe we can still return lists, but we sort them in the unittest function. What do you think?

Reply:
No, sorting here or elsewhere is fine when we're dealing with tests. Alternatively, if we go full sets, `assertSetEqual` is fine. I just meant to say that this particular snippet isn't making use of the features of a set over a list.

```
before_set = set(before_dict.keys())
after_set = set(after_dict.keys())

removed_keys = before_set.difference(after_set)
added_keys = after_set.difference(before_set)

for key in before_set.intersection(after_set)):
  if before_dict[key] == after_dict[key]:
    unchanged_files.add(key)
  else:
    modified_files.add(key)
```

Perhaps we can do something like this and use assertSetEqual in the test?

Reply:
Thanks again for the feedback @adityasaky ! I have force pushed the branch. Please let me know if you feel something is not right.